### DISCUSSION TITLE
starlette 0.27.x does not support anyio 4

### DISCUSSION THREAD

**Discussion by iwanbolzern** [fastapi/fastapi#10187] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI

app = FastAPI()

@app.get("/")
def state() -> Any:
  raise Exception('Any exception')
```

### Description

Hi all, 

Thank you for your great work on fastapi!

We're facing the issue right now that with the release of anyio 4 yesterday and the current dependency tree of:

```
fastapi 0.103.0 FastAPI framework, high performance, easy to learn, fast to code, ready for production
├── pydantic >=1.7.4,<1.8 || >1.8,<1.8.1 || >1.8.1,<2.0.0 || >2.0.0,<2.0.1 || >2.0.1,<2.1.0 || >2.1.0,<3.0.0
│   ├── ...
├── starlette >=0.27.0,<0.28.0
│   └── anyio >=3.4.0,<5
│       ├── idna >=2.8
│       └── sniffio >=1.1
└── typing-extensions >=4.5.0
```

where `starlette 0.27.x` has `anyio <5` as dependency but only `starlette==0.31.0` officially supports anyio 4.

The error we get is:
```
During handling of the above exception, another exception occurred:

  + Exception Group Traceback (most recent call last):
  |   File "C:\Projects\xxx\common\error.py", line 212, in dispatch
  |     response = await call_next(request)
  |                ^^^^^^^^^^^^^^^^^^^^^^^^
  |   File "C:\Projects\xxx\.venv\Lib\site-packages\starlette\middleware\base.py", line 84, in call_next
  |     raise app_exc
  |   File "C:\Projects\xxx\.venv\Lib\site-packages\starlette\middleware\base.py", line 70, in coro
  |     await self.app(scope, receive_or_disconnect, send_no_error)
  |   File "C:\Projects\xxx\.venv\Lib\site-packages\starlette\middleware\base.py", line 106, in __call__
  |     async with anyio.create_task_group() as task_group:
  |   File "C:\Projects\xxx\.venv\Lib\site-packages\anyio\_backends\_asyncio.py", line 664, in __aexit__
  |     raise BaseExceptionGroup(
  | ExceptionGroup: unhandled errors in a TaskGroup (1 sub-exception)
  +-+---------------- 1 ----------------
```

### Operating System

Windows

### FastAPI Version

0.103.0

### Pydantic Version

2.3.0

### Python Version

3.11

**Comment by tiangolo** [comment]
It's very strange, I haven't been able to reproduce this issue here, I don't get the error. I tried manually uninstalling and re-installing/upgrading AnyIO to make sure I had the same versions as you... I'm trying to think what might be happening. Because if I can't reproduce it, I wouldn't have a way to be sure when it's actually solved. 🤔

**Reply by agronholm** [reply]
Nevertheless, the Starlette test client at least tries to import `anyio.start_blocking_portal` which was already moved to `anyio.from_thread.blocking_portal` back in v3.0 but w/o a deprecation message. In v4.0 it was removed from the old location, causing the import to fail.

**Reply by agronholm** [reply]
I need to verify this. I was told this by a user and haven't run into it myself. A closer examination indicates that my previous statement may not actually apply to Starlette 0.27.0 at all.

**Reply by agronholm** [reply]
Indeed, the user I was talking to had a FastAPI version that pinned Starlette 0.24.0 which is indeed even less compatible.

It would still be preferable to get the dependency upgraded to 0.31, as this version of Starlette contains changes to ensure compatibility, such as avoiding deprecated calls and turning unwrapping single-exception groups (thus avoiding the problem the OP is seeing).

**Reply by iwanbolzern** [reply]
Hi @tiangolo and @agronholm, thank you for looking into it!

I went through our application again (where we get the error) and realized that the error appears in combination with a middleware. Please find an updated example attached:

```example.py
from http.client import HTTPException
from typing import Callable

from fastapi import FastAPI
from starlette.middleware.base import BaseHTTPMiddleware


class TestMiddleware(BaseHTTPMiddleware):
    async def dispatch(self, request, call_next: Callable):
        return await call_next(request)


app = FastAPI()

app.add_middleware(TestMiddleware)


@app.get("/")
async def state():
    raise HTTPException('Any exception')
```

```test_example.py
from fastapi.testclient import TestClient
from starlette.status import HTTP_500_INTERNAL_SERVER_ERROR

from example import app


def test_with_anyio_4():
    test_client = TestClient(app)

    res = test_client.get("/")

    assert res.status_code == HTTP_500_INTERNAL_SERVER_ERROR, \
        f'Status code {res.status_code} not equal HTTP_500_INTERNAL_SERVER_ERROR'
```

I hope this helps reproducing the error.

**Reply by agronholm** [reply]
Yes, I made changes in the latest Starlette to fix this specific issue, which is an artifact of how task groups in both asyncio and AnyIO 4 (but not AnyIO 3) work.

**Reply by tiangolo** [reply, marked as answer]
Thanks @iwanbolzern for the example! 🤓 

And thanks @agronholm for the input. 🙇 

For now, as a quick fix, I pinned AnyIO in FastAPI, then I'll upgrade Starlette (I have to do several things for that), and then I'll unpin AnyIO.

The fix is available in FastAPI 0.103.1 🎉

**Reply by agronholm** [reply]
What's the procedure of updating FastAPI to use the latest Starlette? It would be nice to be able to use FastAPI together with AnyIO 4 given the massive improvements in the latter.
